### PR TITLE
Side exercises for raindrops: part1

### PR DESCRIPTION
**Pull request by glennj**
5 exercises here. Part 2 will have 4 more.

| exercise | canonical data | difficulty |
| --- | --- | :-: |
| [accumulate](https://github.com/glennj/tcl/blob/side-exercises-for-hamming/exercises/accumulate/README.md) | no canonical data | 5 |
| [alphametics](https://github.com/glennj/tcl/blob/side-exercises-for-hamming/exercises/alphametics/README.md) | [canonical-data.json](https://github.com/exercism/problem-specifications/blob/master/exercises/alphametics/canonical-data.json) | 6 |
| [binary-search](https://github.com/glennj/tcl/blob/side-exercises-for-hamming/exercises/binary-search/README.md) | [canonical-data.json](https://github.com/exercism/problem-specifications/blob/master/exercises/binary-search/canonical-data.json) | 4 |
| [bob](https://github.com/glennj/tcl/blob/side-exercises-for-hamming/exercises/bob/README.md) | [canonical-data.json](https://github.com/exercism/problem-specifications/blob/master/exercises/bob/canonical-data.json) | 3 |
| [matching-brackets](https://github.com/glennj/tcl/blob/side-exercises-for-hamming/exercises/matching-brackets/README.md) | [canonical-data.json](https://github.com/exercism/problem-specifications/blob/master/exercises/matching-brackets/canonical-data.json) | 2 |

My alphametics implementations is a bit brute force and the test suite takes about 65 seconds on my machine.

I added a hints file for the matching-brackets exercise. It's maybe too wordy?

### CODE REVIEW

File: lib/testMatchers.tcl [hunk 1]
```diff
@@ -27,23 +27,23 @@ proc dictionaryMatch {expected actual} {
...
+#   expr {$a == $b}      ;# => 0
+#   expr {$a && $b}      ;# => 0
+#   expr {!!$a == !!$b}  ;# => 1
+#
 proc booleanMatch {expected actual} {
     return [expr {
         [string is boolean -strict $expected] &&
         [string is boolean -strict $actual] &&
-        (($expected && $actual) || (!$expected && !$actual))
+        !!$expected == !!$actual
```
Comment (sshine) [inline lib/testMatchers.tcl]: Good point.

This apparently leaves the instances of `booleanMatch` in

- exercises/isogram/isogram.test
- exercises/armstrong-numbers/armstrong-numbers.test
- exercises/clock/clock.test

Let's handle those in a separate PR.

**Comment by glennj** [comment]
re: alphametics
> One path for avoiding this burden and keeping the running time low is to convert the example by looking at the [python example](https://github.com/exercism/python/blob/master/exercises/alphametics/example.py). It should contain enough explanation to make this translation more mechanical.

I've been working at translating that python, and I'm down to the last line
```python
        # Chain two iterators
        # first iterator with all non-0 permutations
        # second iterator with all permulations without 1 letter
        # insert 0 in all possible positions of that permutation
        return chain(permutations(nzdigset, totcnt),
                     map(lambda x: x[0][:x[1]] + (0,) + x[0][x[1]:],
                         product(permutations(nzdigset, totcnt - 1),
                                 poslst)))
```
gulp.

Anyway, I'll create an issue for this and merge the PR.

File: exercises/accumulate/accumulate.test [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+    test accumulate-5 "composition" -body {
+        accumulate \
+            pair \
+            [accumulate n $numbers {divmod $n 7}] \
+            {
+                lassign $pair quotient remainder
+                expr {7 * $quotient + $remainder}
+            }
+    } -returnCodes ok -result $numbers
```
[the remark this answers is not in the document]
Comment (glennj) [reply exercises/accumulate/accumulate.test]: done

File: exercises/bob/example.tcl [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+    # Tcl's default boolean result is 1/0 for true/false.
+    # Using that to combine them as a string:
+    switch -exact -- "${silent}${asking}${yelling}" {
+        "100" {return "Fine. Be that way!"}
+        "011" {return "Calm down, I know what I'm doing!"}
+        "010" {return "Sure."}
+        "001" {return "Whoa, chill out!"}
+        "000" {return "Whatever."}
+    }
```
[the remark this answers is not in the document]
Comment (glennj) [reply exercises/bob/example.tcl]: tweaked a bit

File: lib/testMatchers.tcl (hunk 1 above)
Comment (glennj) [reply lib/testMatchers.tcl]: issue created
